src/snowflake/cli/api/commands/flags.py

```diff
@@ -499,70 +501,86 @@ def execution_identifier_argument(sf_object: str, example: str) -> typer.Argumen
     )
 
 
-def project_type_option(project_name: str):
-    from snowflake.cli.api.exceptions import NoProjectDefinitionError
-    from snowflake.cli.api.project.definition_manager import DefinitionManager
+def register_project_definition(project_name: Optional[str], is_optional: bool) -> None:
+    project_path = cli_context_manager.project_path_arg
+    env_overrides_args = cli_context_manager.project_env_overrides_args
 
-    def _callback(project_path: Optional[str]):
-        dm = DefinitionManager(project_path)
+    try:
+        dm = DefinitionManager(project_path, {CONTEXT_KEY: {"env": env_overrides_args}})
         project_definition = dm.project_definition
         project_root = dm.project_root
+        template_context = dm.template_context
 
-        if not getattr(project_definition, project_name, None):
+        if project_name is not None and not getattr(
+            project_definition, project_name, None
+        ):
             raise NoProjectDefinitionError(
                 project_type=project_name, project_file=project_path
             )
 
-        cli_context_manager.set_project_definition(project_definition)
-        cli_context_manager.set_project_root(project_root)
-        cli_context_manager.set_template_context(dm.template_context)
-        return project_definition
-
-    if project_name == "native_app":
-        project_name_help = "Snowflake Native App"
-    elif project_name == "streamlit":
-        project_name_help = "Streamlit app"
+    except MissingConfiguration:
+        if is_optional:
+            project_definition = None
+            project_root = None
+            template_context = {
+                CONTEXT_KEY: {
+                    "env": ProjectEnvironment(
+                        default_env={}, override_env=env_overrides_args
+                    )
+                }
+            }
+        else:
+            raise
```

Comment:
Can we make this using if's instead of raising and handling exception?

Reply:
yes - refactored.